Create Association Rules for the Market Basket Analysis for the given Threshold

In [ ]:

!pip install apyori

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5956 sha256=678315175f933a8850149a00fb9faedb71503e7672df66da063ac63f591378a9
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [ ]:
import pandas as pd
import numpy as np
from apyori import apriori

In [ ]:

from google.colab import drive

drive.mount('/content/gdrive/', force_remount=True)

Mounted at /content/gdrive/


In [ ]:
store_data = pd.read_csv("/content/gdrive/MyDrive/Datasets/GroceryStoreDataSet.csv",encoding='latin-1')

In [ ]:
store_data.head()

MILK,BREAD,BISCUIT
0  BREAD,MILK,BISCUIT,CORNFLAKES
1            BREAD,TEA,BOURNVITA
2           JAM,MAGGI,BREAD,MILK
3              MAGGI,TEA,BISCUIT
4            BREAD,TEA,BOURNVITA

In [ ]:
store_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19 entries, 0 to 18
Data columns (total 1 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   MILK,BREAD,BISCUIT  19 non-null     object
dtypes: object(1)
memory usage: 280.0+ bytes


In [ ]:
#Renaming the column to PRODUCTS

store_data.rename(columns = {'MILK,BREAD,BISCUIT':'PRODUCTS'}, inplace = True)

In [ ]:
# Creating a list of items in every transaction.
#The list will work as a training set from where we can generate the list of Association Rules.

list_items = list(store_data["PRODUCTS"].apply(lambda x:x.split(',')))
list_items

[['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

In [ ]:
from mlxtend.preprocessing import TransactionEncoder

encode = TransactionEncoder()
encode_data = encode.fit(list_items).transform(list_items)
store_data = pd.DataFrame(encode_data,columns = encode.columns_)
store_data.head()


BISCUIT  BOURNVITA  BREAD   COCK  COFFEE  CORNFLAKES    JAM  MAGGI   MILK  \
0     True      False   True  False   False        True  False  False   True   
1    False       True   True  False   False       False  False  False  False   
2    False      False   True  False   False       False   True   True   True   
3     True      False  False  False   False       False  False   True  False   
4    False       True   True  False   False       False  False  False  False   

   SUGER    TEA  
0  False  False  
1  False   True  
2  False  False  
3  False   True  
4  False   True

In [ ]:
#Applying apriori algorithm
#The lift value of an association rule is the ratio of the confidence of the rule and the expected confidence of the rule.

association_rules = apriori(list_items, min_support=0.0045, min_confidence=0.2, min_lift=3, min_length=2)
association_list = list(association_rules)


In [ ]:
#Generating the association rule between the items

for i in range(0, len(association_list)):
    print(association_list[i][0])

frozenset({'JAM', 'MAGGI'})
frozenset({'CORNFLAKES', 'BISCUIT', 'BREAD'})
frozenset({'COCK', 'BISCUIT', 'COFFEE'})
frozenset({'COCK', 'CORNFLAKES', 'BISCUIT'})
frozenset({'CORNFLAKES', 'BISCUIT', 'COFFEE'})
frozenset({'MILK', 'CORNFLAKES', 'BISCUIT'})
frozenset({'BISCUIT', 'TEA', 'MAGGI'})
frozenset({'BOURNVITA', 'SUGER', 'COFFEE'})
frozenset({'MILK', 'CORNFLAKES', 'BREAD'})
frozenset({'JAM', 'MAGGI', 'BREAD'})
frozenset({'MILK', 'JAM', 'BREAD'})
frozenset({'COCK', 'CORNFLAKES', 'COFFEE'})
frozenset({'MILK', 'CORNFLAKES', 'COFFEE'})
frozenset({'CORNFLAKES', 'COFFEE', 'TEA'})
frozenset({'MILK', 'TEA', 'COFFEE'})
frozenset({'MILK', 'CORNFLAKES', 'TEA'})
frozenset({'MILK', 'JAM', 'MAGGI'})
frozenset({'TEA', 'JAM', 'MAGGI'})
frozenset({'MILK', 'CORNFLAKES', 'BISCUIT', 'BREAD'})
frozenset({'BISCUIT', 'TEA', 'MAGGI', 'BREAD'})
frozenset({'COCK', 'CORNFLAKES', 'BISCUIT', 'COFFEE'})
frozenset({'MILK', 'JAM', 'MAGGI', 'BREAD'})
frozenset({'TEA', 'JAM', 'MAGGI', 'BREAD'})
frozenset({'MILK', 'COR

In [ ]:
#####  APRIORI 2

In [ ]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules
df = pd.read_csv("/content/gdrive/MyDrive/Datasets/GroceryStoreDataSet.csv", names = ['transaction'], sep = ',')

pandas for dataframe management
mlxtend (machine learning extensions) is a Python library of useful tools for day-to-day data science tasks.

mlxtend.preprocessing for data preprocessing before performing the task.

mlxtend.frequent_patterns for association rule task.

load data as a variable df

In [ ]:
#2. One-hot encoding
#We have to change the data to the appropriate format before inputting it into the Apriori algorithm. The goal is a one-hot DataFrame.

#Change the DataFrame to a list of lists.

df = list(df["transaction"].apply(lambda x:x.split(",")))


In [ ]:
#df = list(df["transaction"].apply(lambda x:x.split(",")))

TypeError: ignored

In [ ]:
#Use TransactionEncoder from the mlxtend library to change the list of transactions to a one-hot array.

In [ ]:
one_hot_transformer = TransactionEncoder()
df_transform = one_hot_transformer.fit_transform(df)

In [ ]:
#Change the one-hot array to a DataFrame.
df = pd.DataFrame(df_transform,columns=one_hot_transformer.columns_)

This is our desired format, one-hot. The columns are items in the store and each row represents a transaction. If the value is True, that item is sold in that transaction. Now, the data is ready to be fed to the algorithm

In [ ]:
#3. Find the frequent itemsets using Apriori

df = apriori(df, min_support = 0.2, use_colnames = True)
df.sort_values(['support'],ascending=False, inplace = True)

 the minimum support to 0.2, so it will filter only a set of item(s) whose support is greater than 0.2 (or more than 20% from 19 transactions).

Only these itemsets which are considered important will proceed to the association rule.

In [ ]:
#4. Association rule

df_ar = association_rules(df, metric="lift", min_threshold=1)
df_ar

antecedents   consequents  antecedent support  consequent support  \
0        (MILK)       (BREAD)                0.25                0.65   
1       (BREAD)        (MILK)                0.65                0.25   
2       (SUGER)       (BREAD)                0.30                0.65   
3       (BREAD)       (SUGER)                0.65                0.30   
4  (CORNFLAKES)      (COFFEE)                0.30                0.40   
5      (COFFEE)  (CORNFLAKES)                0.40                0.30   
6      (COFFEE)       (SUGER)                0.40                0.30   
7       (SUGER)      (COFFEE)                0.30                0.40   
8         (TEA)       (MAGGI)                0.35                0.25   
9       (MAGGI)         (TEA)                0.25                0.35   

   support  confidence      lift  leverage  conviction  
0      0.2    0.800000  1.230769    0.0375    1.750000  
1      0.2    0.307692  1.230769    0.0375    1.083333  
2      0.2    0.666667  1.025641    0.0050    1.050000  
3      0.2    0.307692  1.025641    0.0050    1.011111  
4      0.2    0.666667  1.666667    0.0800    1.800000  
5      0.2    0.500000  1.666667    0.0800    1.400000  
6      0.2    0.500000  1.666667    0.0800    1.400000  
7      0.2    0.666667  1.666667    0.0800    1.800000  
8      0.2    0.571429  2.285714    0.1125    1.750000  
9      0.2    0.800000  2.285714    0.1125    3.250000

The association_rules function will automatically calculate key metrics of our transaction data including support, confidence, lift, leverage, and conviction.

The result is required interpretation for further business actions.

Interpretation examples

From the highest confidence in index 0, the confidence is 0.8. It means that the customer who buys milk will buy bread for 80%. However, you have to keep in mind that confidence is not everything. The high confidence in this row is due to the high support of bread (consequence support 0.65) which means bread occurs in many transactions, so it will not make a business impact if we try to sell bread with milk anyway.
The better metric is lift or leverage. Index 8 has the highest lift and leverage at 2.28 and 0.1125 respectively. It means that the customers who buy tea are likely to buy Maggi as well.
What we can do after knowing this?

Change shelf layout to sell Maggi close to tea.
Promoting Maggi together with tea.